File: data-structures-greedy-algorithms-complete/greedy_algorithms.py
```python
import heapq

def fractional_knapsack(capacity, items):
    """
    Fractional Knapsack using Greedy approach.

    Args:
        capacity: Knapsack weight capacity.
        items: List of (value, weight) tuples.

    Returns:
        Total value of items in the knapsack.
    """
    items.sort(key=lambda x: x[0] / x[1], reverse=True)  # Sort by value-to-weight ratio
    total_value = 0
    remaining_capacity = capacity
    for value, weight in items:
        if weight <= remaining_capacity:
            total_value += value
            remaining_capacity -= weight
        else:
            fraction = remaining_capacity / weight
            total_value += value * fraction
            break
    return total_value

# Add other algorithms (Huffman Coding, Activity Selection, Coin Change) here...


def huffman_coding(frequencies):
    """Huffman Coding Algorithm"""
    heap = [(freq, char) for char, freq in frequencies.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        lo = heapq.heappop(heap)
        hi = heapq.heappop(heap)
        for pair in lo[1:]:
            pair[0] = pair[0] + hi[0]
        heapq.heappush(heap,(lo[0] + hi[0],lo[1:] + hi[1:]))

    return heap[0]


def activity_selection(activities):
    """Activity Selection Problem"""
    activities.sort(key=lambda x: x[1])  # Sort by finish time
    selected_activities = []
    last_finish_time = 0
    for start, finish in activities:
        if start >= last_finish_time:
            selected_activities.append((start, finish))
            last_finish_time = finish
    return selected_activities

def min_coins(amount, coins):
    """Coin Change (Minimum Coins) using Greedy approach."""
    coins.sort(reverse=True)  # Sort coins in descending order
    num_coins = 0
    remaining_amount = amount
    for coin in coins:
        while remaining_amount >= coin:
            remaining_amount -= coin
            num_coins += 1
    if remaining_amount == 0:
        return num_coins
    else:
        return -1 # No solution found


```

File: leetcode-greedy-algorithms-solutions-complete/greedy_algorithms.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include "utils.h" // Include helper functions and data structures


using namespace std;

// Fractional Knapsack (Greedy)
double fractionalKnapsack(vector<Item>& items, int capacity) {
    // Sort items by value-to-weight ratio
    sort(items.begin(), items.end(), [](const Item& a, const Item& b) {
        return (double)a.value / a.weight > (double)b.value / b.weight;
    });

    double totalValue = 0;
    int remainingCapacity = capacity;

    for (const Item& item : items) {
        if (item.weight <= remainingCapacity) {
            totalValue += item.value;
            remainingCapacity -= item.weight;
        } else {
            totalValue += (double)item.value / item.weight * remainingCapacity;
            remainingCapacity = 0;
            break;
        }
    }
    return totalValue;
}


//Activity Selection (Greedy)
int activitySelection(vector<Activity>& activities) {
    //Sort activities by finish time
    sort(activities.begin(), activities.end(), [](const Activity& a, const Activity& b){
        return a.finish < b.finish;
    });

    int count = 0;
    int lastFinishTime = 0;
    for(const auto& activity : activities){
        if(activity.start >= lastFinishTime){
            count++;
            lastFinishTime = activity.finish;
        }
    }
    return count;
}


// ... Implement Huffman Coding and Coin Change (with both greedy and potentially DP solutions) ...

int main() {
    // Example usage (Fractional Knapsack)
    vector<Item> items = {{10, 60}, {20, 100}, {30, 120}};
    int capacity = 50;
    cout << "Fractional Knapsack Max Value: " << fractionalKnapsack(items, capacity) << endl;


    //Example Usage (Activity Selection)
    vector<Activity> activities = {{1,4},{3,5},{0,6},{5,7},{3,9},{5,9},{6,10},{8,11},{8,12},{2,14},{12,16}};
    cout << "Number of Non-overlapping Activities: " << activitySelection(activities) << endl;

    // ... Add example usage for Huffman Coding and Coin Change ...

    return 0;
}